vector/src/main/java/im/vector/app/features/settings/devices/v2/ParseDeviceUserAgentUseCase.kt

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2022 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.settings.devices.v2
+
+import im.vector.app.features.settings.devices.v2.list.DeviceType
+import org.matrix.android.sdk.api.extensions.orFalse
+import javax.inject.Inject
+
+class ParseDeviceUserAgentUseCase @Inject constructor() {
+
+    fun execute(userAgent: String?): DeviceUserAgent {
+        if (userAgent == null) return createUnknownUserAgent()
+
+        return when {
+            userAgent.contains(ANDROID_KEYWORD) -> parseAndroidUserAgent(userAgent)
+            userAgent.contains(IOS_KEYWORD) -> parseIosUserAgent(userAgent)
+            userAgent.contains(DESKTOP_KEYWORD) -> parseDesktopUserAgent(userAgent)
+            userAgent.contains(WEB_KEYWORD) -> parseWebUserAgent(userAgent)
+            else -> createUnknownUserAgent()
+        }
+    }
+
+    private fun parseAndroidUserAgent(userAgent: String): DeviceUserAgent {
+        val appName = userAgent.substringBefore("/")
+        val appVersion = userAgent.substringAfter("/").substringBefore(" (")
+        val deviceInfoSegments = userAgent.substringAfter("(").substringBeforeLast(")").split("; ")
+        val deviceModel: String?
+        val deviceOperatingSystem: String?
+        if (deviceInfoSegments.firstOrNull() == "Linux") {
+            val deviceOperatingSystemIndex = deviceInfoSegments.indexOfFirst { it.startsWith("Android") }
+            deviceOperatingSystem = deviceInfoSegments.getOrNull(deviceOperatingSystemIndex)
+            deviceModel = deviceInfoSegments.getOrNull(deviceOperatingSystemIndex + 1)
+        } else {
+            deviceModel = deviceInfoSegments.getOrNull(0)
+            deviceOperatingSystem = deviceInfoSegments.getOrNull(1)
+        }
+        return DeviceUserAgent(DeviceType.MOBILE, deviceModel, deviceOperatingSystem, appName, appVersion)
+    }
+
+    private fun parseIosUserAgent(userAgent: String): DeviceUserAgent {
+        val appName = userAgent.substringBefore("/")
+        val appVersion = userAgent.substringAfter("/").substringBefore(" (")
+        val deviceInfoSegments = userAgent.substringAfter("(").substringBeforeLast(")").split("; ")
+        val deviceModel = deviceInfoSegments.getOrNull(0)
+        val deviceOperatingSystem = deviceInfoSegments.getOrNull(1)
+        return DeviceUserAgent(DeviceType.MOBILE, deviceModel, deviceOperatingSystem, appName, appVersion)
```

Comment:
We do not distinguish iOS and Android devices? I thought the idea would be to render them differently in the UI.

Reply:
Yes, the design suggests only one icon for mobile devices. Instead, we will render device model info in session details.